Implementations for Merge, IDataView and ReadCsv

pgovind opened the pull request:
Should go in after https://github.com/dotnet/corefxlab/pull/2687

More work towards https://github.com/dotnet/corefx/issues/26845

pgovind commented on src/Microsoft.Data/BaseColumn.cs:
```diff
@@ -59,18 +60,20 @@ protected set
             get => GetValue(startIndex, length);
         }
 
+        public virtual void Resize(long length) => throw new NotImplementedException();
```
Add a comment about why this is needed

pgovind commented on src/Microsoft.Data/DataFrame.IO.cs:
```diff
@@ -0,0 +1,261 @@
+            foreach (var line in read)
+            {
+                if (col >= line.Length)
+                    throw new FormatException(string.Format(Strings.LessColumnsThatExpected, nbline + 1)); // Will this really work?
```
The question is w.r.t localization

eerhardt replied:
Yes, you can have placeholders in resources `{0}`.

---
In reply to: [299212735](https://github.com/dotnet/corefxlab/pull/2692#discussion_r299212735) [](ancestors = 299212735)

pgovind commented on src/Microsoft.Data/DataFrame.IO.cs:
```diff
@@ -0,0 +1,261 @@
+            for (int i = 0; i < numCol; ++i)
+            {
+                Type kind = GuessKind(i, lines);
+                switch (kind)
```
Support exists for now to create the different types of columns, but `GuessKind` itself only returns one of `string, bool, float` at the moment.

pgovind commented on src/Microsoft.Data/DataFrame.cs:
```diff
@@ -260,6 +261,305 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
+                // Create empty columns
+                BaseColumn column = other.Column(i).Clone(emptyMap);
+                int index = ret._table.GetColumnIndex(column.Name);
+                if (index != -1)
```
Refactor to use the SetSuffix... method

pgovind commented on src/Microsoft.Data/StringColumn.cs:
```diff
@@ -45,6 +45,17 @@ public StringColumn(string name, IEnumerable<string> values) : base(name, 0, typ
+            }
+        }
+
         public void Append(string value)
```
Implementation of `HashColumnValues` for StringColumn is coming in the next patch

eerhardt commented:
@pgovind - can you rebase or merge this with the current master? I don't want to re-review the Join code.

pgovind commented:
> @pgovind - can you rebase or merge this with the current master? I don't want to re-review the Join code.

Done

eerhardt commented on src/Microsoft.Data/DataFrame.IO.cs:
```diff
@@ -0,0 +1,261 @@
+        /// <param name="encoding">text encoding</param>
+        /// <param name="index">add one column with the row index</param>
+        /// <returns>DataFrame</returns>
+        public static DataFrame ReadCsv(string filename,
```
I almost wonder if these methods should be external to the `DataFrame`. That way we can ensure anyone can take any data (like from SQL, or elsewhere) and materialize a DataFrame.

pgovind replied:
By external, do you mean an API such as 

`ReadCsv(filename, out DataFrame dataFrame)`?

eerhardt replied:
I mean not on the `DataFrame` class. The reasoning is that other code outside of the current assembly should be able to load `DataFrame` objects from different sources - SQL, Parquet, in-memory objects, etc. If `ReadCsv` is on the `DataFrame` class itself, it has access to `private` fields/methods that other code won't be able to use. This means it gets `special` treatment. If someone else wanted to write a "create a DataFrame from some data source" method, would they be able to do it?

eerhardt commented on src/Microsoft.Data/DataFrame.IO.cs:
```diff
@@ -0,0 +1,261 @@
+        {
+            if (a == typeof(string) || b == typeof(string))
+                return typeof(string);
+            if (a == typeof(float) || b == typeof(float))
```
What about `double` or `int`, etc?

pgovind replied:
Right, `MaxKind` only has 1 caller `DetermineType` and `DetermineKind` always passes in one of string, float or bool. Or do you mean to add support now to make it easier to parse `double`/`int` in the future?

eerhardt replied:
Maybe the question is - why aren't we supporting `double` and `int` types? Why just `float`?

pgovind replied:
Ah, that's because we only read in the first 10 lines in the csv file and try to guess the types from it. Since the first 10 lines of column A could be `int`, but everything else after that a `float`, I chose `float` as the common numeric type. At some point, it'd probably be worth it to have a strongly typed API that specifies the column types

eerhardt commented on src/Microsoft.Data/Microsoft.Data.DataFrame.csproj:
```diff
@@ -18,6 +18,11 @@
     <PackageReference Include="System.Memory" Version="4.5.1" />
     <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="4.5.1" />
     <PackageReference Include="System.Text.Encoding" Version="4.3.0" />
+    <PackageReference Include="Microsoft.ML" Version="1.1.0" />
```
Don't reference the whole `Microsoft.ML`.  But instead, `Microsoft.ML.DataView`.

I wonder if that functionality should be split into a separate assembly. That way the core DataFrame library doesn't need to depend on Microsoft.ML.

eerhardt commented on src/Microsoft.Data/Microsoft.Data.DataFrame.csproj:
```diff
@@ -18,6 +18,11 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Microsoft.Experimental.Collections\Microsoft.Experimental.Collections.csproj" />
```
Do we really need to be dependent on this experimental library? Is it planning on shipping?

/cc @safern

safern replied:
I think he depends on MultiValueDictionary... the plan for experimental collections is that if a collection gets popular enough and might be useful for partners or the framework, we will move that collection into the product.

danmosemsft replied:
I think MultiValueDictionary already got formally API reviewed (as part of Ian's internship). That was before my time, but assuming so then most likely it would only need a brief re-review. And I think it should likely go into CoreFX for 5.0. It seems a fairly basic, uncontroversial type.

eerhardt replied:
Is the plan for MultiValueDictionary to support `netstandard2.0` or lower?

eerhardt replied:
IMHO - for this case, using `Dictionary<TKey, ICollection<long>>` (or similar) would suffice, and it would alleviate the need for an added dependency. Making `DataFrame` dependent on a non-stable library affects how and when we could ship it in the future.

pgovind replied:
Ah, I was worried this would be the case. I can change to using a `Dictionary`.

eerhardt commented on tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs:
```diff
@@ -744,5 +744,68 @@ public void TestJoin()
             VerifyJoin(join, left, right, JoinAlgorithm.Inner);
         }
+
+        [Fact]
```
We need tests for ReadCsv , IDataView.

eerhardt commented on tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs:
```diff
@@ -744,5 +744,68 @@ public void TestJoin()
+            DataFrame merge = left.Merge<int>(right, "Int", "Int");
+            Assert.Equal(10, merge.RowCount);
+            Assert.Equal(merge.ColumnCount, left.ColumnCount + right.ColumnCount);
+            Assert.Null(merge["Int_right"][6]);
```
We should be validating the whole resultant dataframe's values.

eerhardt commented on src/Microsoft.Data/StringColumn.cs:
```diff
@@ -45,6 +45,17 @@ public StringColumn(string name, IEnumerable<string> values) : base(name, 0, typ
+
+            for (long i = Length; i < length; i++)
+            {
+                Append(null);
```
Is doing this one at a time the best approach?

eerhardt commented on src/Microsoft.Data/DataFrameTable.cs:
```diff
@@ -147,5 +147,82 @@ public int GetColumnIndex(string columnName)
+                {
+                    case Type boolType when boolType == typeof(bool):
+                        parse = bool.TryParse(value, out bool boolAppend);
+                        (column as PrimitiveColumn<bool>).Append(parse ? boolAppend : default);
```
It feels wrong to me that the DataFrameTable is responsible for parsing these values. It feels more tied to the CSV reading implementation.

eerhardt commented on src/Microsoft.Data/DataFrame.IO.cs:
```diff
@@ -0,0 +1,261 @@
+                              guess_rows: guess_rows, index: index);
+        }
+
+        private delegate StreamReader FunctionCreateStreamReader();
```
Why not just use a `Func<StreamReader>`?

eerhardt commented on src/Microsoft.Data/DataFrame.IO.cs:
```diff
@@ -0,0 +1,261 @@
+        private delegate StreamReader FunctionCreateStreamReader();
+
+        /// <summary>
+        /// Reads a text file as a IDataView.
```
`as a IDataView`?

pgovind replied:
Good catch. Left over from scratch work I was doing

eerhardt commented on src/Microsoft.Data/DataFrame.IO.cs:
```diff
@@ -0,0 +1,261 @@
+                var val = line[col];
+                try
+                {
+                    bool.Parse(val);
```
`TryParse` instead of `Parse` and an exception. Throwing exceptions in happy paths will cause perf degradation.

eerhardt replied:
Are you planning on profiling the CSV parsing and seeing where any bottlenecks are?

---
In reply to: [299702122](https://github.com/dotnet/corefxlab/pull/2692#discussion_r299702122) [](ancestors = 299702122)

pgovind replied:
Yup. Sometime soon. There's a whole bunch of perf tests I want to do.

eerhardt commented on src/Microsoft.Data/DataFrame.IO.cs:
```diff
@@ -0,0 +1,261 @@
+            }
+
+            // Fills values.
+            using (var st = createStream())
```
We already have ALL the values in memory, right? Why read them from the stream again?

pgovind replied:
No, we don't. `lines` only has a length of `guess_rows` (defaults to 10 rows)

eerhardt commented on src/Microsoft.Data/DataFrame.IO.cs:
```diff
@@ -0,0 +1,261 @@
+
+            if (index)
+            {
+                // TODO: Bug in InsertColumn. Can only insert at the end. Should support inserting in the middle
```
What's the plan for handling this TODO?

pgovind replied:
Filed https://github.com/dotnet/corefxlab/issues/2695

pgovind commented:
Considering the comments, this will likely take me a bit of time to refactor. I'm going to close this PR for now and make 3 separate ones addressing the comments for Merge, IDataView and ReadCsv respectively.
